### PR TITLE
video recording for evaluation environments using tensorboard

### PR DESCRIPTION
**Pull request by Bam4d**
Added an option for generating videos during evaluation runs:
`--eval-capture-videos`

also allowed customization of number processes used for evaluation, so we are a bit more flexible there.
`--eval-processes`

uses `writer.add_video` from tensorboard.

Videos end up in wandb looking like this and are linked to specific eval runs by global_step variable: 
image

### CODE REVIEW

**Comment by Bam4d** [comment]
closes #152

File: enn_ppo/enn_ppo/train.py [hunk 1]
```diff
@@ -336,6 +342,14 @@ def run(
                 self.actions.extend(action)
                 self.logprobs.extend(logprob)
 
+            if capture_videos:
+                self.rendered = np.concatenate(
```
Comment (cswinter) [inline enn_ppo/enn_ppo/train.py:346]: This looks like it will take O(n^2) time

File: enn_ppo/enn_ppo/train.py [hunk 2]
```diff
@@ -102,6 +100,10 @@ def parse_args(override_args: Optional[List[str]] = None) -> argparse.Namespace:
         help='number of parallel environments in eval')
     parser.add_argument('--eval-env-kwargs', type=str, default=None,
         help='JSON dictionary with keyword arguments for the eval environment')
+    parser.add_argument('--eval-processes', type=int, default=1,
```
Comment (cswinter) [inline enn_ppo/enn_ppo/train.py:103]: Let's make the default `None` and default to `--processes` in that case.

File: entity_gym/entity_gym/environment/env_list.py [hunk 3]
```diff
@@ -34,6 +36,9 @@ def reset(self, obs_space: ObsSpace) -> ObsBatch:
             self.cls.action_space(),
         )
 
+    def render(self, **kwargs: Mapping[str, Any]) -> np.ndarray:
```
Comment (cswinter) [inline entity_gym/entity_gym/environment/env_list.py:39]: I think we probably want a more generic render method that doesn't assume a specific rendering format, but this seems fine for now. We should still add some docs here to describe the encoding used by the output array, and use `numpy.typing.NDArray` for more precise type information (see e.g. train.py).

**Review by cswinter — APPROVED** [review]
LGTM just have a few nits.

File: entity_gym/entity_gym/environment/env_list.py (hunk 3 above)
Comment (Bam4d) [reply entity_gym/entity_gym/environment/env_list.py:39]: So I'm specifically following the format that add_video allows. I did not end up using the Griddly render tool (which uses ffmpeg and mp4). I think by following tensorboard's format we're probably as close to "generic render method" as we can get.

File: enn_ppo/enn_ppo/train.py (hunk 1 above)
Comment (Bam4d) [reply enn_ppo/enn_ppo/train.py:346]: I didnt think concatenate was O(n)... I can change this to do a single np.concatenate at the end?

Comment (cswinter) [reply enn_ppo/enn_ppo/train.py:346]: I don't actually know how concatenate is implemented, but my first assumption would have been it needs to allocate a new array and copy the two source arrays. Single concatenate at the end should work.

**Pull request merged**